DDDEastAnglia/Helpers/LoginMethods/ExternalLoginsProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using DDDEastAnglia.Helpers.Security;
+using DDDEastAnglia.Models;
+using Microsoft.Web.WebPages.OAuth;
+
+namespace DDDEastAnglia.Helpers.LoginMethods
+{
+    public class ExternalLoginsProvider
+    {
+        private readonly IOAuthClientInfo oAuthClientInfo;
+
+        public ExternalLoginsProvider(IOAuthClientInfo oAuthClientInfo)
+        {
+            if (oAuthClientInfo == null)
+            {
+                throw new ArgumentNullException("oAuthClientInfo");
+            }
+            
+            this.oAuthClientInfo = oAuthClientInfo;
+        }
+
+        public IEnumerable<LoginMethod> GetForUser(string userName)
+        {
+            var accounts = oAuthClientInfo.GetAccountsFromUserName(userName);
+            var externalLogins = new List<LoginMethod>();
+
+            foreach (OAuthAccount account in accounts)
+            {
+                var clientData = oAuthClientInfo.GetOAuthClientData(account.Provider);
+
+                externalLogins.Add(new LoginMethod
+                {
+                    DisplayName = clientData.DisplayName,
+                    ProviderName = account.Provider,
+                    ProviderUserId = account.ProviderUserId
+                });
+            }
+
+            return externalLogins.OrderBy(l => l.DisplayName);
+        }
+
+        public IEnumerable<LoginMethod> GetAllAvailable()
+        {
+            var authenticationClientDatas = oAuthClientInfo.RegisteredClientData();
+            var externalLogins = new List<LoginMethod>();
+
+            foreach (var externalLogin in authenticationClientDatas)
```

Comment:
Seems like there's some duplicated code between these methods. Is it possible to pull out a helper method to reduce that? [Not necessarily, I know; the code does look just enough different, but perhaps it can be munged to make it work.]